wgpu_shadertoy/imgui.py

```diff
@@ -0,0 +1,227 @@
+        {new_line.join(var_init_lines)}
+    }};
+    layout(binding = {constant_binding_idx}) uniform ConstantInput{constant_binding_idx} const_input{constant_binding_idx};
+    {new_line.join(var_mapping_lines)}
+    """
+    # the identifier name includes the digit so common doesn't cause redefinition!
+    # TODO messed up indentation... textwrap.dedent?
+    return code_construct
+
+def replace_constants(code: str, constants: list[ShaderConstant], constant_binding_idx: int) -> str:
+    """
+    comment out existing constants and redefine them with uniform struct
+    """
+    code_lines = code.splitlines()
+    for const in constants:
+        # comment out existing constants
+        code_lines[const.line_number] = f"// {code_lines[const.line_number]}"
+
+    constant_headers = construct_imports(constants, constant_binding_idx)
+    code_lines.insert(0, constant_headers)
+
+    return "\n".join(code_lines)
+
+
+# imgui stuff
+def update_gui():
+    # todo: look at exmaples nad largely copy nad paste, will be called in the draw_frame function I think.
+
+    pass
+
+
+def gui(renderpasses: list["RenderPass"]):
+    ig.new_frame()
+    ig.set_next_window_pos((0, 0), ig.Cond_.appearing)
+    ig.set_next_window_size((0, 0), ig.Cond_.appearing) # auto size not wide enough with text :/
+    ig.begin("Shader constants", None)
+    ig.text('in-dev imgui overlay\n')
+
+    if ig.is_item_hovered():
+        ig.set_tooltip("TODO")
+
+    # maybe we should have a global main or utils.get_main()?
+    main = renderpasses[0].main
+
+    # TODO: avoid duplication, maybe common should be a renderpass instance (at least a little bit) - or we iterate through constants lists
+    if main._common_constants:
+        if ig.collapsing_header("Common Constants", flags=ig.TreeNodeFlags_.default_open):
+            for const in main._common_constants:
+                if const.shader_dtype == "float":
+                    _, main._common_constants_data[const.name] = ig.drag_float(f"{const.name}", main._common_constants_data[const.name], v_speed=abs(const.value)*0.01)
+                elif const.shader_dtype == "int":
+                    _, main._common_constants_data[const.name] = ig.drag_int(f"{const.name}", main._common_constants_data[const.name], v_speed=abs(const.value)*0.01)
+                if ig.is_item_hovered() and ig.is_mouse_clicked(ig.MouseButton_.right):
+                    main._common_constants_data[const.name] = const.value
+
+    for rp in renderpasses: # TODO: most likely add common here?
+        constants = rp._constants
+        constants_data = rp._constants_data
+        if ig.collapsing_header(f"{rp} Constants", flags=ig.TreeNodeFlags_.default_open):
+            if hasattr(rp, "texture_front"): # isinstance(rp, BufferRenderPass)
+                # make this another toggle? or a whole 2nd UI?
+                front_view = rp.texture_front.create_view()
+                front_ref = rp.main._imgui_backend.register_texture(front_view)
```

Comment:
Hi, came cross here from https://github.com/pygfx/wgpu-py/issues/729#issuecomment-3172852887.
It’s generally better not to create and register new texture views every frame in the render loop. Instead, a global texture view should be registered and used during initialization, or a imgui.TextureRef can be registered and created only at the first time the GUI is rendered.
If it’s absolutely necessary to create and register new texture views each frame in the render loop, you must manually call the newly added `unregister` method in https://github.com/pygfx/wgpu-py/pull/749 to release the old texture view resources.